### DISCUSSION TITLE
Automated GraphQL SDK

### DISCUSSION THREAD

**Discussion by jeffsee55** [tinacms/tinacms#2249] [Experimental]
We've released a new experimental feature which auto-generates an "SDK" for data fetching (thanks @logan-anderson 🎉 ). We've heard from a lot of developers who've said that it can be a bit of a pain to jump back and forth between the `.tina/schema.ts` and their respective queries to ensure they're querying for all of the fields when making schema changes. We're hoping that an auto-generated solution will solve most of those issues while also giving better typescript-safety. Please give it a shot and let us know what you think!

```
yarn add @tinacms/cli@latest
```

For detailed instructions read the docs [here](https://github.com/tinacms/tinacms/blob/main/packages/%40tinacms/cli/GeneratedClientDocs.md). 

You can see it implemented in our [example starter](https://github.com/tinacms/tinacms/blob/main/experimental-examples/tina-cloud-starter/pages/%5Bfilename%5D.tsx#L12) for further reference.

**Comment by jamespohalloran** [comment]
This is looking great!

I just had a few gotchas when looking through the examples:

## Client function casing

In the [example](https://github.com/tinacms/tinacms/blob/main/experimental-examples/tina-cloud-starter/pages/%5Bfilename%5D.tsx#L13), there's a few functions that are uppercase on the client (e.g `client.ContentQuery` ). I'm not sure if that should be a matter of documenting queryNames like:

```
query contentQuery($relativePath: String!) { { /**/ }
```

or if we should update the function names on the client to make them start lowercase. 

## Verbose Query/client response

- The naming around the generated response gets a bit verbose.

e.g
```
  const pageList = await client.getPagesList();
  return {
    paths: pageList.data.getPagesList.edges.map((page) => ({
      params: { filename: page.node.sys.filename },
    })),
    fallback: false,
  };
```

`pageList.data.getPagesList.edges.`

Thismay be more scoped to the GraphQL API rather than the client, but could we remove "get" from the query/response response in "getPagesList" (Most GraphQL APIs have seen, your graphql query/response would be "pageList" instead of "getPageList").

Do "edges" and "node" need to be apart of the response?

## Nested queries per collection

 - This probably would increase scope, but has there been any thought on having the client nest its functions within each collection?
 E.g
 
 ```
 const postResponse = await client.posts.get("foo-bar.md")
 ```

 ```
 const postResponse = await client.posts.list()
 ```
 
It's probably not a super big deal, but I think it would make the autocomplete easier by nesting each of the functions in the collection.

**Reply by jeffsee55** [reply]
The more opinionated SDK (version C mentioned [here](https://github.com/tinacms/product/discussions/80)) needs to handle a lot of the type generation ourselves so will be a good opportunity to take on stronger opinions like this, and likely make it feel a bit less like graphql. We'll be able to remove the `get{Something}Document` etc, which I would be a fan of - I like the suggestion of nesting them in `post.{something}`. So @logan-anderson to your point, rather than a chaining approach it might be easier to more closely match the shape of a GraphQL query. The API I was able to actually get working looked like this:
```js
const response = await client(sdk => {
  myPost: sdk.post("hello.md")
  postList: sdk.postList({filter: {category: "ACTION"}, first: 10, include: {director: true})
})
```
Which just maps to this query:
```graphql
query {
  myPost: getPostDocument(relativePath: "home.mdx") {
    ...
  }
  postList: getPostList {
    ...
  }
}
```
To me this seems like it'd be the easiest to implement and the result would look like this:
```js
{
  myPost: {
    title: "Hello"
  },
  postList: {
    edges: {
      node: {
        title: 'Hello',
        director: {
          __typename: 'director',
          name: 'George'
        }
      }
    }
  }
}
```

I could see an argument for paginate: false
```js
const response = await client(sdk => {
  myPost: sdk.post("hello.md")
  postList: sdk.postList({filter: {category: "ACTION"}, first: 10, include: {director: true}, paginate: false)
})

{
  myPost: {
    title: "Hello"
  },
  postList: {
    title: 'Hello',
    director: {
      __typename: 'director',
      name: 'George'
    }
  }
}
```

**Comment by logan-anderson** [comment]
@jamespohalloran 
> Client function casing

Good catch ill update the example.
> Verbose Query/client response

Hostly, yeah this is very true but its sorta the nature of graphql. We _could_ return  just the list of posts but what would we do in the case of a query like this? when we want to better support pagination.

For example: 
```graphql
query Pages {
  getPostsList {
    pageInfo {
      hasPreviousPage
      hasNextPage
      startCursor
      endCursor
    }
    totalCount
    edges {
      cursor
      node {
        id
        sys {
          filename
        }
        data {
          _body
          title
          date
        }
      }
    }
  }
}
```

> Nested queries per collection

```ts
const postResponse = await client.posts.get("foo-bar.md")
```

This is a really good idea and a much better API. Just might be hard to support when the user wants to fetch two content types at once.

```ts
// one network request 
const postResponse = await client.posts.get("foo-bar.md")

// Second network request
const global = await client.global.get("index.json")

// now we would have 2 seperate queries so we would have this issue

// (we would have to write a stitch function)
return {
  props: {
      stitch(global.query,  postResponse.query)
  }
}
```

When really the user should only have to make one graphql request and not two. We might be able to fix it with something like this.

```ts
// one network request
const response = await client.posts.list().global.get('index.json').execute()

// one query
return {
  props: {
     query: response.query
  }
}
```

**Comment by Perni1984** [comment]
We really welcome this feature, but are running into a "problem" with the generated fragments for our "real-world" schema, leaving us with a broken type generation for now.

We are receiving a lot of errors because of fields that have the same names, but different types / constraints - e.g. there can be an `imageSrc` prop in the one component that is required and a `imageSrc` prop in another component that is not required.

```
AggregateError: GraphQL Document Validation failed with 52 errors;
  Error 0: GraphQLDocumentError: Fields "imageSrc" conflict because they return conflicting types "String!" and "String". Use different aliases on the fields to fetch both if this was intentional.
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:8:7
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:37:7

Error 1: GraphQLDocumentError: Fields "imageSrc" conflict because they return conflicting types "String!" and "String". Use different aliases on the fields to fetch both if this was intentional.
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:8:7
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:202:7

Error 2: GraphQLDocumentError: Fields "imageSrc" conflict because they return conflicting types "String" and "String!". Use different aliases on the fields to fetch both if this was intentional.
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:37:7
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:209:7

Error 3: GraphQLDocumentError: Fields "imageSrc" conflict because they return conflicting types "String" and "String!". Use different aliases on the fields to fetch both if this was intentional.
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:37:7
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:291:7

Error 4: GraphQLDocumentError: Fields "imageSrc" conflict because they return conflicting types "String" and "String!". Use different aliases on the fields to fetch both if this was intentional.
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:202:7
    at /home/perni/projects/pcode-web-next/.tina/__generated__/frags.gql:209:7
```

Is there already a way to specify aliases in the schema? Because when we wrote the queries ourselves it was easy for us to specify the alias "by hand".

I also pointed out that behaviour before in https://discord.com/channels/835168149439643678/835172431836479508/900492050905055303

Is it possible to disable the automated generation of the types and restore the old behaviour?

**Reply by jeffsee55** [reply]
This is a great point, we'll have to think about this a bit more and in the meantime will allow the SDK generation to be turned on/off with the CLI. Thanks @Perni1984!

**Reply by cereallarceny** [reply]
@Perni1984 I'm actually encountering this very problem now. Super frustrating where I'll have a key deeply nested in an object for one template and it will conflict with the generated type of another key on some other template.

@jeffsee55 Do you all have any current plans to revisit schema generation? I've hit this problem about 6 times so far in trying out TinaCMS - sometimes with required fields, sometimes with the naming of a field that happens to have the same name somewhere else. It always results in the types failing to be generated.

**Reply by jeffsee55** [reply]
@cereallarceny are those conflicts happening on different levels? My understanding is that the conflicts only occur when the properties are on the _same_ level:

```graphql
query {
  page(relativePath: "some-path.md") {
    blocks {
      ...on PageHeroBlock {
        title # required
      }
      ...on PageCtaBlock {
        title # not required
      }
    }
  }
}
```

**Reply by Perni1984** [reply]
@jeffsee55 Unfortunately I cannot remember if this happened on same or different levels.

@cereallarceny we are handling this right now with not requiring any field at all in tina and thus letting our components handle this - .e.g in the worst case scenario we just render nothing or a placeholder component displaying a notice that XYZ would be required to render a component. Not ideal, but works for our use case.

**Comment by Perni1984** [comment]
@jeffsee55 @jamespohalloran @logan-anderson FYI, just wanted to point out that we now started to use the experimental client again for some content pages, but then unfortunately we ran into this error again - so we are forced to step back again to write our own queries.

**Reply by logan-anderson** [reply]
Thanks for trying to use the experimental client! unfortunately because of [this issue with graphQL](https://github.com/graphql/graphql-js/issues/1361) it looks like its not possible to generate the types. Would it be possible to rename some of your image fields to `optionalImgeSrc` and `ImgSrc` (matching to the type) ? This would avoid the name conflicts. We are also open to other ideas as well, so feel free to let us know if you have any ideas how to fix it 😄 .

**Comment by logan-anderson** [comment]
# The next iteration of the automated graphQL client.

We have identified two key issues with the current graphQL client.

1. Lack of ability to resolve references in data.
    - Right now if you wish to resolve references in your data you will have to write your own queries in the `.tina/queries/` folder.
2. Lack of ability to query more than one piece of content with a single request.
     - Again to do this one would have to write there own queries


We plan on solving these issues as while providing excellent DX and typescript types with the next iteration of the client. 

## What it might look like.

First, it might be able to fetch data as well as return the query and the variables used.

Here are a couple of options of what it might look like. 

### Idea 1

```ts
 const data = await request({
    blogs: (client) => client.getPostsDocument({ relativePath: "anotherPost.md" }),
    pages: (client) => client.getPagesList(),
  });
```

### Idea 2

```ts
const data = await request{
    getAuthorsList: {
      variables: {},
      options: { include: ["posts"] },
    },
    getPostsDocument: {
      variables: { relativePath: "anotherPost.md" },
      options: { include: ["authors"] },
    },
    getGlobalDocument: { variables: { relativePath: "anotherPost.md" } },
  });
```

### Idea 3

```ts
const data = await request({
    getAuthorsList: {
      variables: {},
      options: {
        include: {
          posts: {
            authors: true,
          },
        },
      },
    },
    getPostsDocument: {
      variables: { relativePath: "anotherPost.md" },
      options: {
        include: {
          posts: true,
        },
      },
    },
  });
```

Personally,  I believe that option 2 or 3 are the most viable to implement while providing a really good DX.  

Option 2 takes inspiration [from TypeORM](https://typeorm.io/#/find-options/basic-options) as you provide a list of relations and nested relations are provided by using `.` in strings. We could also type this by using [template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html).

Option 3 takes [inspiration from prisma](https://www.prisma.io/client) where they use object notion to resolve references.

**Reply by jeffsee55** [reply]
The reason I suggested the callback approach in option 1 is because the keys of those top-level objects need to be flexible. In GraphQL, you can do aliases, and it's a really important bit of functionality that we need to keep intact. So in effect, the key should be meaningless to us:

```ts
const data = await request({
    // `myBlogPage` can be whatever you want it to be
    myBlogPage: (client) => client.getPostsDocument({ relativePath: "anotherPost.md" }),
    pages: (client) => client.getPagesList(),
});
```
Maps to:
```graphql
query {
  myBlogPage: getPostsDocument(...) {
    ...
  }
  pages: getPostList {...}
}
```
Aside from that the _values_ side from option 1 can be whatever. I'm a fan of option 3 because it feels way more flexible to me:
```ts
const data = await request({
  getPostsDocument: {
    variables: { relativePath: "anotherPost.md" },
    options: {
      include: {
        authors: true,
      },
    },
  },
});

// Nested included references
const data = await request({
  getPostsDocument: {
    variables: { relativePath: "anotherPost.md" },
    options: {
      include: {
        authors: {
          // instead of a boolean, we can pass a nested config to include 
          // the author's posts, for example.
          include: {
            posts: true
          }
        }
      },
    },
  },
});


// You can narrow the return fields
const data = await request({
  getPostsDocument: {
    variables: { relativePath: "anotherPost.md" },
    options: {
      fields: {
        data: {
          title: true
        }
      }
    },
  },
});

// Or jump back into query-land at any level
const data = await request({
  getPostsDocument: {
    variables: { relativePath: "anotherPost.md" },
    options: {
      fields: {
        data: {
          blocks: `#graphql
            __typename
            ...on CtaBlock {
              title
            }
          `
        }
      }
    },
  },
});
```
I'm confident this can all be well-typed, even the nested query strings.

---

Another thing we haven't discussed is control components that are nested further away. If you had a block that stores a reference you want to include, it might be a bit wordy to have to nest all the way down, instead, I think we could offer a flattened `includes` shape, too:

```ts
const data = await request({
  getPostsDocument: {
    variables: { relativePath: "anotherPost.md" },
    options: {
      include: {
        "blocks.cta.author": true
      }
    }  
  },
});
```
This would be especially handy for mdx references, if we are to ever implement [this suggestion](https://github.com/tinacms/tinacms/discussions/2429#discussioncomment-1971430).

**Comment by jamespohalloran** [comment]
Since we've done a few other DX API updates since last discussing this feature, I just want to circle back on how this could look for an ultra-simple example, with our current API:

```jsx
// pages/_app.js
import TinaCMS from 'tinacms'

const App = ({ Component, pageProps }) => {
  return (
    <TinaCMS
      apiURL="http://localhost:4001/graphql"
    >
      <Component {...pageProps} />
    </TinaCMS>
  )
}

export default App
```
^ Nothing data related happening here, just a Tina wrapper

```jsx
// home.jsx
import React from 'react'
import { useTina } from 'tinacms/dist/edit-state'
import { client } from '../.tina/__generated__/types'

export default function Page(props) {
	const {data} = useTina(props.post)
	return ( <h1>{data.getPostDocument.data.title}</h1>)
}

export const getStaticProps() {
	const postResponse = await client.post.get("hello-world.md")
	return { 
	     props: { 
	          post: postResponse
		}
	}
}
```

I just mean to call attention to how readable this feels, for a full working Tina example (apart from the schema in /.tina) compared to our query-based example in our [docs](https://tina.io/docs/tinacms-context). 

--
(Don't focus on the `client.post.get` call, as I know its implementation is still up for debate). Also, I'm aware it's not code-split here as-is, but could look this way with some of the points brought up in [this discussion](https://github.com/tinacms/tinacms/discussions/2402)

**Reply by jeffsee55** [reply]
@jamespohalloran are you suggesting that the `client.post.get` is the thing you want to introduce as different from what we currently have? I'm not sure I see a significant difference from this compared to what's currently available:
```ts
// Use the props returned by get static props
export default function BlogPostPage(props) {
  const { data } = useTina(props)
  return <Post {...data.getPostsDocument} />;
}

export const getStaticProps = async ({ params }) => {
  const client = ExperimentalGetTinaClient();
  const tinaProps = await client.BlogPostQuery({
    relativePath: `${params.filename}.mdx`,
  });
  return {
    props: {
      ...tinaProps,
    },
  };
};
```
Both examples being limited for brevity, you can see the full example [here](https://github.com/tinacms/tina-cloud-starter/blob/main/pages/posts/%5Bfilename%5D.tsx)(aside from the missing `useTina` hook we haven't added yet)

`BlogPostQuery` is custom, because it also includes a query for nav, but `getPostsDocument` comes [out-of-the-box](https://github.com/tinacms/tina-cloud-starter/blob/main/.tina/__generated__/queries.gql#L1-L16).

**Reply by jamespohalloran** [reply]
> @jamespohalloran are you suggesting that the client.post.get is the thing you want to introduce as different from what we currently have? 

No, as I said: "(Don't focus on the `client.post.get` call, as I know its implementation is still up for debate)" :D

I just mean to resurface this pitch in how it at a high level looks now after some recent API changes (mainly useTina), and that the overall feel of using the client feels good to me (comparing it to the query handling that we do in the [docs' happy path](https://tina.io/docs/tinacms-context/#adding-tina-to-the-sites-frontend) )

> I'm not sure I see a significant difference from this compared to what's currently available:

I'm not pitching a particular client direction here, but moreso just throwing this out there as fodder for, should we either move the SDK out of experimental? or bump up priority on investigating its next steps?

**Reply by jeffsee55** [reply]
Oh I see, ok cool. Probably my main hesitation with this is that you'll run into [this](https://github.com/tinacms/tinacms/discussions/2249#discussioncomment-1988664). Perhaps it's worthwhile to try and think of a solution for that in the short term.

That being said, we could support the ability to generate the SDK and not necessarily generate the base queries, so @Perni1984's issue would at least be avoidable rather than a total blocker. Thoughts @logan-anderson ?

**Comment by solidevolution** [comment]
I'm on my evaluation for upcoming projects and Tina CMS looks amazing. But I'm a little bit confused about some query features. I especially searching for a solution for paging with sorting. In the starter project the blog pages are sorted by filename, but usually you want to sort them by date. Of course it's possible to do this after the GraphQL query in javascript but what about paging?

**Comment by cereallarceny** [comment]
Unfortunately I also have this problem when types are conflicting. It’s not
just with a field being required. If a field is a string in one place and a
number in another, there will be a conflict.

My only thought, and I haven’t confirmed this, would be to wrap all fields
of a template in a uniquely-named top level object field. That’s clearly a
workaround, and I haven’t tested it, but I figure that would allow the type
generator to see a difference.
